File: data structure and algorithms/Bit Manipulation/473.matchsticks-to-square.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <random>
#include <regex>
#include <set>
#include <sstream>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

class Solution {
    int side;
    vector<int> sticks;
    array<int, 4> sides{};

    bool dfs(int i) {
        if (i == (int)sticks.size()) return sides[0] == side && sides[1] == side && sides[2] == side && sides[3] == side;
        int len = sticks[i];
        unordered_set<int> seen;
        for (int j = 0; j < 4; ++j) {
            if (sides[j] + len > side || seen.count(sides[j])) continue;
            seen.insert(sides[j]);
            sides[j] += len;
            if (dfs(i + 1)) return true;
            sides[j] -= len;
        }
        return false;
    }

public:
    bool makesquare(vector<int>& matchsticks) {
        int total = accumulate(matchsticks.begin(), matchsticks.end(), 0);
        if (matchsticks.size() < 4 || total % 4) return false;
        side = total / 4;
        sticks = matchsticks;
        sort(sticks.rbegin(), sticks.rend());
        if (sticks[0] > side) return false;
        sides = {0, 0, 0, 0};
        return dfs(0);
    }
};

/*
Interview explanation:
Backtracking assigns each stick to one of four equal-length sides. Sorting descending makes large restrictive sticks fail early, and skipping side sums already tried avoids symmetric duplicate states.

C++ data structures: array<int,4> is ideal for the four side sums; unordered_set<int> per level removes equivalent placements.

Edge cases: total must be divisible by four and no stick can exceed the side length.

Complexity: worst-case O(4^n), heavily pruned; recursion depth O(n).
*/

File: pt-2/473.matchsticks-to-square.java
import java.util.*;

/**
 * Algorithm:
 * Backtrack by assigning the longest remaining matchstick to one of four side
 * sums. Sorting descending prunes early. Equal side sums are skipped at each
 * level to avoid symmetric duplicate states.
 *
 * Java data structures:
 * int[4] stores the partial side lengths; HashSet<Integer> skips duplicate
 * side states in the same recursion frame.
 *
 * Complexity:
 * Worst-case exponential, bounded by O(4^n), with O(n) recursion space.
 */
class Solution {
    private int[] matchsticks;
    private int[] sides;
    private int side;

    public boolean makesquare(int[] matchsticks) {
        int total = 0;
        for (int x : matchsticks) {
            total += x;
        }
        if (matchsticks.length < 4 || total % 4 != 0) {
            return false;
        }
        side = total / 4;
        Arrays.sort(matchsticks);
        reverse(matchsticks);
        if (matchsticks[0] > side) {
            return false;
        }
        this.matchsticks = matchsticks;
        sides = new int[4];
        return dfs(0);
    }

    private boolean dfs(int i) {
        if (i == matchsticks.length) {
            return sides[0] == side && sides[1] == side && sides[2] == side && sides[3] == side;
        }
        int length = matchsticks[i];
        Set<Integer> seen = new HashSet<>();
        for (int j = 0; j < 4; j++) {
            if (seen.contains(sides[j]) || sides[j] + length > side) {
                continue;
            }
            seen.add(sides[j]);
            sides[j] += length;
            if (dfs(i + 1)) {
                return true;
            }
            sides[j] -= length;
        }
        return false;
    }

    private void reverse(int[] a) {
        for (int l = 0, r = a.length - 1; l < r; l++, r--) {
            int tmp = a[l];
            a[l] = a[r];
            a[r] = tmp;
        }
    }
}

